src/storage/src/source/mod.rs

```diff
@@ -478,9 +527,15 @@ pub struct SourceMessage<Key, Value> {
     /// Headers, if the source is configured to pass them along. If it is, but there are none, it
     /// passes `Some([])`
     pub headers: Option<Vec<(String, Option<Vec<u8>>)>>,
+
+    /// Allow sources to optionally output a specific differential
+    /// `diff` value. Defaults to `+1`.
```

Comment:
I'm curious why you've structured it as an arbitrary `Diff` type rather than either:
- Option<mz_repr::Diff>
- `mz_repr::Diff` and having everything other than postgres set it to 1 (since they can authoritatively say that)

I don't follow (perhaps out of ignorance!) when we'd need to support a non-`mz_repr::Diff` in this field?

Reply:
I had it `Option<mz_repr::Diff>` or something similar but I found that it made it unclear in `render_source` what was happening. with the associated type we can ensure that ONLY
`SourceType::Row` ends up having `type Diff = Diff`, where as other types always have `type Diff = ()` .

in the future this also enables us to generic-ize `DecodeResult` and cleanup `render_source` even more

Reply:
so in short: yes, we want other diff types, namely `()` to clearly communicate when a source wants to use the default behavior, as opposed to all those sources needing to implement that default themselves

Reply:
with a fixed type, either `diff: i64` or `diff: Option<i64>`, the `SourceReader` impl is required to provide the "default" to us. If kafka's `SourceReader` gives us `2` over `None` or `1`, we can't really tell if thats wrong or not, we have to assume its right

With this, we declare, at the top of the trait impl, that the kafka source reader can produce only _1_ value as a diff (in this case `()`, but we could make an empty struct call `Default` or whatever, preventing any complexity or thought in the decision when producing the `SourceMessage` inside the `next` method.

I found this to be much cleaner and easier to reason about, and less fragile. You are right that it doesn't buy us much, but I am inclined to encode invariants in the type system as much as possible, without making stuff unreadable. I found this to be more readable than the previous version (you may be able to find a commit on this pr from earlier) that didn't do it this way

Reply:
> If kafka's SourceReader gives us 2 over None or 1, we can't really tell if thats wrong or not, we have to assume its right

Isn't `KafkaSourceReader` the authoritative source for what is correct for a kakfa source though? Just like the `PostgresSourceReader` is the authoritative source for the diff field for a postgres source.  We're not supposing what we should protect against postgres being wrong.  I think that's the thing I'm not understanding.

Hmm I guess this setup is analogous to parametrizing over a trait implemented by two enums: `enum AppendOnly { Append }` and `enum AppendOrDelete { Append, Delete }` -- which I do very clearly see the value in soperhaps it's the naming tripping me up? I think writing those enums out is more explicit (and maybe even more clear but I was the one who came up with it so grain of salt...) but if you don't want to do that, spitballing here: what about just renaming `specific_diff -> update_type`?

Reply:
@cjubb39 `KafkaSourceReader` isnt really the authoritative source: it can be composed with envelopes that may or may not support non-`1` diffs; so in some sense, the caller in `render::sources::render_source` is authoritative (as it should be, it is the place where `SourceReaders` are correctly instantiated

we can definitely change the name! `update_type` is fine, what about `type MessageDiff` and `message_diff`?